src/engine/strat_engine/cmd.rs

```diff
@@ -0,0 +1,155 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Handles invoking external binaries.
+// This module assumes that, for a given machine, there is only one place
+// where the desired executable might be installed. It expects the engine
+// to identify that place at its initialization by invoking verify_binaries(),
+// and to exit immediately if verify_binaries() return an error. If this
+// protocol is followed then when any command is executed the unique absolute
+// path of the binary for this machine will already have been identified.
+// However stratisd may run for a while and it is possible for the binary
+// to be caused to be uninstalled while stratisd is being run. Therefore,
+// the existence of the file is checked before the command is invoked, and
+// an explicit error is returned if the executable can not be found.
+
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+use uuid::Uuid;
+
+use stratis::{ErrorEnum, StratisError, StratisResult};
+
+/// Find the binary with the given name by looking in likely locations.
+/// Return None if no binary was found.
+fn find_binary(name: &str) -> Option<PathBuf> {
+    ["/usr/sbin", "/sbin", "/usr/bin", "/bin"]
+        .iter()
+        .map(|pre| [pre, name].iter().collect::<PathBuf>())
+        .find(|path| path.exists())
+}
+
+// These are the external binaries that stratisd relies on.
+// Any change in this list requires a corresponding change to BINARIES,
+// and vice-versa.
+const MKFS_XFS: &str = "mkfs.xfs";
+const THIN_CHECK: &str = "thin_check";
+const THIN_REPAIR: &str = "thin_repair";
+const XFS_ADMIN: &str = "xfs_admin";
+const XFS_GROWFS: &str = "xfs_growfs";
+
+lazy_static! {
+    static ref BINARIES: HashMap<String, Option<PathBuf>> = [
+        (MKFS_XFS.to_string(), find_binary(MKFS_XFS)),
+        (THIN_CHECK.to_string(), find_binary(THIN_CHECK)),
+        (THIN_REPAIR.to_string(), find_binary(THIN_REPAIR)),
+        (XFS_ADMIN.to_string(), find_binary(XFS_ADMIN)),
+        (XFS_GROWFS.to_string(), find_binary(XFS_GROWFS))
+    ].iter()
+        .cloned()
+        .collect();
+}
+
+/// Verify that all binaries that the engine might invoke are available at some
+/// path. Return an error if any are missing. Required to be called on engine
+/// initialization.
+pub fn verify_binaries() -> StratisResult<()> {
+    // Would like to use find() here instead of loop, but compiler bug
+    // https://github.com/rust-lang/rust/issues/51415 makes that impossible.
+    for (ref name, ref path) in BINARIES.iter() {
+        if path.is_none() {
+            return Err(StratisError::Engine(ErrorEnum::NotFound, name.to_string()));
+        }
+    }
+    Ok(())
+}
+
+/// Common function to call a command line utility, returning an Result with an error message which
+/// also includes stdout & stderr if it fails.
+fn execute_cmd(cmd: &mut Command, error_msg: &str) -> StratisResult<()> {
+    let result = cmd.output()?;
```

Comment:
I think is would be better to remove the `exists` check from `get_executable` and place it in the error path for `cmd.output()` so that we can than either raise the missing file error or return the error returned from trying to execute the command.

Reply:
My point was that instead of having to `stat` the executable every time to see if it exists before it's called we could just assume that the executable is still there, which it likely is and only if we encounter an error trying to execute it, see if it had gone missing.  Thus the error message would be more helpful.  The end result should be the same, the difference is when the existence check is done, either every time an executable is called or only when we encounter an error.

Reply:
This would require some significant alterations to existing code for very moderate gain. I don't think it's worth doing unless this PR is used as a test case for the failure crate.

Reply:
The suggested change makes the code behave more like the usual exception handling model, which would be an improvement. However, I would still like to merge this as is, and then experiment with the failure crate which I hope will allow me to implement the suggestion in a better manner than could be managed at this time.